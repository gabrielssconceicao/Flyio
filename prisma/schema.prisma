// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// Looking for ways to speed up your queries, or scale easily with your serverless or edge functions?
// Try Prisma Accelerate: https://pris.ly/cli/accelerate-init

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id         String   @id @default(uuid())
  name       String
  username   String   @unique
  email      String   @unique
  password   String
  bio        String?
  profileImg String?
  active     Boolean  @default(true)
  createdAt  DateTime @default(now())
  updatedAt  DateTime @updatedAt

  posts     Post[]
  PostLikes PostLikes[]
}

model Post {
  id        String       @id @default(uuid())
  content   String
  createdAt DateTime     @default(now())
  updatedAt DateTime     @updatedAt
  userId    String
  user      User         @relation(fields: [userId], references: [id])
  images    PostImages[]
  PostLikes PostLikes[]
}

model PostImages {
  id        String   @id @default(uuid())
  url       String
  postId    String
  createdAt DateTime @default(now())
  post      Post     @relation(fields: [postId], references: [id], onDelete: Cascade)
}

model PostLikes {
  postId String
  userId String

  user User @relation(fields: [userId], references: [id])
  post Post @relation(fields: [postId], references: [id], onDelete: Cascade)

  @@id([postId, userId])
}
