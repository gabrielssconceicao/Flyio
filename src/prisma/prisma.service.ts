import { Injectable, OnModuleInit } from '@nestjs/common';
import { Prisma, PrismaClient } from '@prisma/client';

@Injectable()
export class PrismaService extends PrismaClient implements OnModuleInit {
  async onModuleInit() {
    await this.$connect();
  }

  async onModuleDestroy() {
    await this.$disconnect();
  }

  async findAll<T>(
    model: {
      count: (args?: Prisma.Args<any, 'count'>) => Promise<number>;
      findMany: (args?: Prisma.Args<any, 'findMany'>) => Promise<T[]>;
    },
    options?: {
      where?: any;
      select?: any;
      orderBy?: any;
      skip?: number;
      take?: number;
    },
  ) {
    const { where, select, orderBy, skip, take } = options || {};

    const [count, items] = await Promise.all([
      model.count({ where }),
      model.findMany({ where, select, orderBy, skip, take }),
    ]);
    return { count, items };
  }
}
