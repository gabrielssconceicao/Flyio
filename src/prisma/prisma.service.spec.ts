import { Test, TestingModule } from '@nestjs/testing';
import { PrismaService } from './prisma.service';

describe('<PrismaService />', () => {
  let service: PrismaService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [PrismaService],
    }).compile();

    service = module.get<PrismaService>(PrismaService);
    jest.spyOn(service, '$connect').mockResolvedValue(undefined);
    jest.spyOn(service, '$disconnect').mockResolvedValue(undefined);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });
  it('should call $connect on module init', async () => {
    const connectSpy = jest.spyOn(service, '$connect');
    await service.onModuleInit();
    expect(connectSpy).toHaveBeenCalledTimes(1);
  });

  it('should call $disconnect on module destroy', async () => {
    const disconnectSpy = jest.spyOn(service, '$disconnect');
    await service.onModuleDestroy();
    expect(disconnectSpy).toHaveBeenCalledTimes(1);
  });

  describe('findAll', () => {
    it('should return count and items', async () => {
      const items = [];
      const count = 0;
      const mockModel = {
        count: jest.fn().mockResolvedValue(count),
        findMany: jest.fn().mockResolvedValue(items),
      };

      const options = {
        where: {},
        select: {},
        orderBy: {},
        skip: 0,
        take: 10,
      };

      const result = await service.findAll(mockModel, options);

      expect(mockModel.count).toHaveBeenCalledWith({ where: options.where });
      expect(mockModel.findMany).toHaveBeenCalledWith(options);
      expect(result).toEqual({
        count,
        items,
      });
      expect(result).toMatchSnapshot();
    });
  });
});
