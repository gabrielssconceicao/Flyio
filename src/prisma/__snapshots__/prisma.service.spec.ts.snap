// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PrismaService /> findAll should return count and items 1`] = `
{
  "count": 0,
  "items": [],
}
`;
